File: frontend/lib/passwordValidation.ts
export const passwordRequirements = [
    { label: "At least 10 characters", test: (value: string) => Array.from(value).length >= 10 },
    { label: "At least one uppercase letter (A-Z)", test: (value: string) => /[A-Z]/.test(value) },
    { label: "At least one lowercase letter (a-z)", test: (value: string) => /[a-z]/.test(value) },
    { label: "At least one symbol (e.g. !, @, #)", test: (value: string) => /[\p{P}\p{S}]/u.test(value) },
];

File: backend/app/password_validation.py
"""Shared password policy for registration and password changes."""

import re
import unicodedata


def validate_password(value: str) -> str:
    """Return the password unchanged, or raise ValueError listing unmet rules."""
    errors = []
    if len(value) < 10:
        errors.append("at least 10 characters")
    if not re.search(r"[A-Z]", value):
        errors.append("at least one uppercase letter (A-Z)")
    if not re.search(r"[a-z]", value):
        errors.append("at least one lowercase letter (a-z)")
    if not any(unicodedata.category(char)[0] in "PS" for char in value):
        errors.append("at least one symbol (e.g. !, @, #)")
    if errors:
        raise ValueError("Password must contain " + "; ".join(errors) + ".")
    return value
